### DISCUSSION TITLE
Marketplace will enforce extension keyword limit of 10

### DISCUSSION THREAD

**Discussion by prashantvc** [microsoft/vscode-discussions#426] [Announcements]
**Update:** We turned off the tag validation for now. We apologies for breaking your workflow. Please note that we will enable this feature in few weeks again, I suggest you to keep `keywords`/`tags` under 10

We enabled the validation rule to reject any extensions with more than 10 keywords in the `package.json`.
This is to improve the visual noise in the extension item details page, improve search performance and prevent extension authors from abusing `keyword` field

The rule is enforced on new uploads; existing extensions will continue to show the initial number of tags/keywords.

You will `You exceeded the number of allowed tags of 10.` error during the new/update extension upload if the package exceeds allowed limit

**Comment by leonelsanchesdasilva** [comment]
@prashantvc Do you know how we can specify the tags to use in the `extension.vsixmanifest` through `vsce package` command? Currently I need to open the VSIX file as a zip and edit this file manually.

**Reply by gjsjohnmurray** [reply]
I think you use the `keywords` section of your `package.json`, for example:

```json
  "keywords": [
    "foo",
    "bar",
    "baz"
  ],

```

**Reply by leonelsanchesdasilva** [reply]
@gjsjohnmurray It worked, but also added a lot of keywords I didn't want, like "theme,icon-theme,snippet,debuggers".

**Reply by gjsjohnmurray** [reply]
Do you have a `categories` section?

**Reply by leonelsanchesdasilva** [reply]
I do:

```
    "categories": [
        "Programming Languages"
    ],
```

**Reply by alefragnani** [reply]
Maybe this is the reason behind these _automatic_ keywords https://github.com/microsoft/vscode-vsce/issues/83.

Based on the description of this announcement, I understand the validation will be only in the tags an extension author adds to `package.json`, ignoring these automatic keywords added by `vsce`, right?

**Reply by leonelsanchesdasilva** [reply]
@alefragnani In my case it is adding the keywords + other things. I'm not sure how to disable the others.

**Reply by alefragnani** [reply]
I guess you can’t…

**Reply by prashantvc** [reply]
It's `keywords` field in the package.json

**Comment by PEZ** [comment]
What's the intended use of the keywords? One of my extensions has exactly 10, and I want to clean it up, but realize I don't know how it is used well enough to delete/add the right ones.

**Reply by alefragnani** [reply]
I understand it is another field the search tool when you try to find extensions, just like its name, description, author. Simply because I don’t believe it does parse the README file (which I would like, btw 😄 )

**Comment by PEZ** [comment]
I ran into this limitation now and my extension ([Calva](https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.calva)) does not get published:
* https://app.circleci.com/pipelines/github/BetterThanTomorrow/calva/6065/workflows/0cce3666-0ec5-4781-b935-6112cfb3f57b/jobs/28463

It is a particularly important release, and I don't know what to do about the publish failure. I have 7 items in the `keywords` entry:

```json
  "keywords": [
    "Clojure",
    "ClojureScript",
    "EDN",
    "REPL",
    "Interactive Programming",
    "Structural Editing",
    "CIDER"
  ],
```

Please advice, @prashantvc. 🙏

**Reply by PEZ** [reply]
I checked in the `extension.vsixmanifest` file in my VSIX, and can't see any keywords there, but there is a `<Tags>` element with all this:

```xml
<Tags>Clojure,ClojureScript,EDN,REPL,Interactive Programming,Structural Editing,CIDER,snippet,keybindings,debuggers,clojure,__ext_clj,__ext_cljs,__ext_cljc,__ext_cljd,__ext_cljx,__ext_clojure,__ext_edn,__ext_joke,__ext_boot,__ext_bb,__ext_calvarepl,__sponsor_extension</Tags>
```

If it counts all those, then I really don't know what to do about it...

**Reply by alefragnani** [reply]
I would say I understand the new limitation, but it shouldn't take in consideration those automatic keywords/tags. 
Only the ones that the extension provides.

**Reply by prashantvc** [reply]
This sounds like a bug! I am sorry about the trouble, let us fix this. I am extremely sorry about it

**Comment by serbani** [comment]
My extension has 5 keywords and the Marketplace adds up another 4 automatic keywords but the extension is still rejected wit "ERROR  You exceeded the number of allowed tags of 10.". How is it counting?

**Reply by serbani** [reply]
I was wrong, the Marketplace shows 9 keywords, the vsix extension.vsixmanifest has 39 automatic keywords added by vsce so even if I remove my 5 keywords I still can't publish the extension.

**Reply by PEZ** [reply]
> and the Marketplace adds up another 4 automatic keywords

Where do you see these?

**Reply by PEZ** [reply]
I see them now. **Tags**, right? My published extension have 13 tags there, only 10 of which are specified as `keywords` in the manifest.

**Reply by serbani** [reply]
Yes, that's the one.

For the time being I've manually edited the `<extension>.vsix/extension.vsixmanifest` and removed the extra tags from the `<Tags>...</Tags>` section, then I manually uploaded the vsix in the Marketplace web interface.

**Comment by serbani** [comment]
There seems to be an issue opened on vsce with the same topic:
https://github.com/microsoft/vscode-vsce/issues/815
